File: host/tests/test_fusion.cpp
#include <gtest/gtest.h>
#include <cmath>
#include <random>
#include <vector>

#include "fusion/fusion.hpp"

using namespace imu;

namespace {

struct SyntheticRow {
  double true_pitch_deg;
  double accel_x, accel_y, accel_z;
  double gyro_y;
};

// Same synthetic-motion idea as
// ../../sim/imu_sim/synth.py's generate_motion() -- a device gently
// rocking in pitch, no real linear acceleration, gyro carrying a fixed
// bias -- reimplemented here rather than shared, so this test file has
// no dependency on Python at all.
std::vector<SyntheticRow> generateMotion(double durationS, double sampleRateHz, unsigned seed,
                                          double gyroBiasDegS = 2.0) {
  std::mt19937 rng(seed);
  std::normal_distribution<double> accelNoise(0.0, 0.03);
  std::normal_distribution<double> gyroNoise(0.0, 0.5);

  const double dt = 1.0 / sampleRateHz;
  const int n = static_cast<int>(durationS * sampleRateHz);
  const double amplitudeDeg = 30.0;
  const double omega = 2.0 * 3.14159265358979323846 * 0.1;

  std::vector<SyntheticRow> rows;
  rows.reserve(n);
  for (int i = 0; i < n; ++i) {
    double t = i * dt;
    double truePitch = amplitudeDeg * std::sin(omega * t);
    double trueRate = amplitudeDeg * omega * std::cos(omega * t);
    double pitchRad = truePitch * (3.14159265358979323846 / 180.0);

    SyntheticRow row;
    row.true_pitch_deg = truePitch;
    row.accel_x = accelNoise(rng);
    row.accel_y = std::sin(pitchRad) + accelNoise(rng);
    row.accel_z = std::cos(pitchRad) + accelNoise(rng);
    row.gyro_y = trueRate + gyroBiasDegS + gyroNoise(rng);
    rows.push_back(row);
  }
  return rows;
}

double rmse(const std::vector<double>& a, const std::vector<double>& b) {
  double sum = 0.0;
  for (size_t i = 0; i < a.size(); ++i) {
    double d = a[i] - b[i];
    sum += d * d;
  }
  return std::sqrt(sum / static_cast<double>(a.size()));
}

}  // namespace

TEST(AccelToPitch, LevelDeviceReadsZero) {
  EXPECT_NEAR(accelToPitchDeg(0.0, 1.0, 0.0), 0.0, 1e-6);
}

TEST(AccelToPitch, NinetyDegreeTilt) {
  EXPECT_NEAR(accelToPitchDeg(1.0, 0.0, 0.0), 90.0, 1e-6);
}

TEST(Fusion, ComplementaryFilterBeatsGyroOnlyOverTime) {
  auto rows = generateMotion(20.0, 200.0, 1);
  const double dt = 1.0 / 200.0;

  ComplementaryFilter comp(0.98);
  double gyroOnlyAngle = 0.0;
  std::vector<double> truth, fused, gyroOnly;
  for (const auto& row : rows) {
    double accelAngle = accelToPitchDeg(row.accel_y, row.accel_z, row.accel_x);
    fused.push_back(comp.update(accelAngle, row.gyro_y, dt));
    gyroOnlyAngle += row.gyro_y * dt;
    gyroOnly.push_back(gyroOnlyAngle);
    truth.push_back(row.true_pitch_deg);
  }

  size_t half = truth.size() / 2;
  std::vector<double> truthHalf(truth.begin() + half, truth.end());
  std::vector<double> fusedHalf(fused.begin() + half, fused.end());
  std::vector<double> gyroOnlyHalf(gyroOnly.begin() + half, gyroOnly.end());

  EXPECT_LT(rmse(truthHalf, fusedHalf), rmse(truthHalf, gyroOnlyHalf));
}

TEST(Fusion, KalmanFilterBeatsGyroOnlyOverTime) {
  auto rows = generateMotion(20.0, 200.0, 1);
  const double dt = 1.0 / 200.0;

  KalmanFilter1D kf;
  double gyroOnlyAngle = 0.0;
  std::vector<double> truth, fused, gyroOnly;
  for (const auto& row : rows) {
    double accelAngle = accelToPitchDeg(row.accel_y, row.accel_z, row.accel_x);
    fused.push_back(kf.update(accelAngle, row.gyro_y, dt));
    gyroOnlyAngle += row.gyro_y * dt;
    gyroOnly.push_back(gyroOnlyAngle);
    truth.push_back(row.true_pitch_deg);
  }

  size_t half = truth.size() / 2;
  std::vector<double> truthHalf(truth.begin() + half, truth.end());
  std::vector<double> fusedHalf(fused.begin() + half, fused.end());
  std::vector<double> gyroOnlyHalf(gyroOnly.begin() + half, gyroOnly.end());

  EXPECT_LT(rmse(truthHalf, fusedHalf), rmse(truthHalf, gyroOnlyHalf));
}

TEST(Fusion, KalmanEstimatesTheGyroBias) {
  auto rows = generateMotion(30.0, 200.0, 1, /*gyroBiasDegS=*/2.0);
  const double dt = 1.0 / 200.0;
  KalmanFilter1D kf;
  for (const auto& row : rows) {
    double accelAngle = accelToPitchDeg(row.accel_y, row.accel_z, row.accel_x);
    kf.update(accelAngle, row.gyro_y, dt);
  }
  EXPECT_NEAR(kf.bias(), 2.0, 0.5);
}

File: sim/tests/test_fusion.py
import sys
import math
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))

from imu_sim.fusion import ComplementaryFilter, KalmanFilter1D, accel_to_pitch
from imu_sim.synth import generate_motion


def test_accel_to_pitch_level_device_reads_zero():
    # Level device: all 1g on Z, nothing on X/Y -- should read ~0 degrees.
    assert abs(accel_to_pitch(accel_y=0.0, accel_z=1.0, accel_x=0.0)) < 1e-6


def test_accel_to_pitch_90_degrees_tilt():
    # Tipped 90 degrees onto its side: all 1g now on Y, nothing on Z.
    pitch = accel_to_pitch(accel_y=1.0, accel_z=0.0, accel_x=0.0)
    assert abs(pitch - 90.0) < 1e-6


def _rmse(a, b):
    n = len(a)
    return math.sqrt(sum((x - y) ** 2 for x, y in zip(a, b)) / n)


def test_gyro_only_integration_drifts_away_from_truth():
    """The whole reason fusion exists: naive gyro-only integration drifts
    steadily because of the fixed bias baked into generate_motion's synthetic
    gyro readings -- this test proves the drift is real, not a strawman."""
    rows = generate_motion(duration_s=20.0, sample_rate_hz=200.0, seed=1)
    dt = 1.0 / 200.0

    gyro_only_angle = 0.0
    truth, gyro_only = [], []
    for row in rows:
        gyro_only_angle += row["gyro_y"] * dt
        truth.append(row["true_pitch_deg"])
        gyro_only.append(gyro_only_angle)

    # Error grows over time -- compare the first second's RMSE to the last
    # second's RMSE and confirm the drift is real and growing, not just noise.
    early_error = _rmse(truth[:200], gyro_only[:200])
    late_error = _rmse(truth[-200:], gyro_only[-200:])
    assert late_error > early_error * 3  # drift compounds; this is a conservative margin


def test_complementary_filter_beats_gyro_only_over_time():
    rows = generate_motion(duration_s=20.0, sample_rate_hz=200.0, seed=1)
    dt = 1.0 / 200.0

    comp = ComplementaryFilter(alpha=0.98)
    gyro_only_angle = 0.0
    truth, fused, gyro_only = [], [], []
    for row in rows:
        accel_angle = accel_to_pitch(row["accel_y"], row["accel_z"], row["accel_x"])
        fused.append(comp.update(accel_angle, row["gyro_y"], dt))
        gyro_only_angle += row["gyro_y"] * dt
        gyro_only.append(gyro_only_angle)
        truth.append(row["true_pitch_deg"])

    # Judge on the second half only -- the complementary filter needs a
    # short settling period before its correction term catches up from a
    # cold start at angle=0.
    half = len(truth) // 2
    assert _rmse(truth[half:], fused[half:]) < _rmse(truth[half:], gyro_only[half:])


def test_kalman_filter_beats_gyro_only_over_time():
    rows = generate_motion(duration_s=20.0, sample_rate_hz=200.0, seed=1)
    dt = 1.0 / 200.0

    kf = KalmanFilter1D()
    gyro_only_angle = 0.0
    truth, fused, gyro_only = [], [], []
    for row in rows:
        accel_angle = accel_to_pitch(row["accel_y"], row["accel_z"], row["accel_x"])
        fused.append(kf.update(accel_angle, row["gyro_y"], dt))
        gyro_only_angle += row["gyro_y"] * dt
        gyro_only.append(gyro_only_angle)
        truth.append(row["true_pitch_deg"])

    half = len(truth) // 2
    assert _rmse(truth[half:], fused[half:]) < _rmse(truth[half:], gyro_only[half:])


def test_kalman_estimates_the_gyro_bias():
    # generate_motion's default gyro_bias_deg_s=2.0 -- after enough samples
    # to settle, the Kalman filter's own bias state should land close to it.
    rows = generate_motion(duration_s=30.0, sample_rate_hz=200.0, seed=1)
    dt = 1.0 / 200.0
    kf = KalmanFilter1D()
    for row in rows:
        accel_angle = accel_to_pitch(row["accel_y"], row["accel_z"], row["accel_x"])
        kf.update(accel_angle, row["gyro_y"], dt)
    assert abs(kf.bias - 2.0) < 0.5
